torchdata/dataloader2/utils/worker.py

```diff
@@ -62,9 +65,36 @@ def process_init_fn(
     r"""
     Based on the worker information, shard the ``DataPipe`` graph dynamically.
     """
-    torch.utils.data.graph_settings.apply_sharding(
-        datapipe, worker_info.num_workers, worker_info.worker_id, SHARDING_PRIORITIES.MULTIPROCESSING
-    )
+    # Find if there is non-replicable DataPipe
+    graph = traverse_dps(datapipe)
+    non_replicable_dp = find_dps(graph, _DummyIterDataPipe)  # type: ignore
+
+    # There are two cases for DataPipe graph in terms of mp sharding:
+    # 1) All DataPipes are replicable, apply mp sharding to the whole graph
+    if len(non_replicable_dp) == 0:
+        torch.utils.data.graph_settings.apply_sharding(
+            datapipe, worker_info.num_workers, worker_info.worker_id, SHARDING_PRIORITIES.MULTIPROCESSING
+        )
+    # 2) There is non-replicable DataPipe. Since we have replaced the lowest common
+    #    ancestor by a `_DummyIterDataPipe`, we would only apply mp sharding
+    #    to replicable branches that don't have `_DummyIterDataPipe`.
+    else:
+        assert len(non_replicable_dp) == 1
+        replicable_branches = find_replicable_branches(graph)
+        for dp in replicable_branches:
+            torch.utils.data.graph_settings.apply_sharding(
+                dp, worker_info.num_workers, worker_info.worker_id, SHARDING_PRIORITIES.MULTIPROCESSING
+            )
+
+        req_queue = non_replicable_dp[0].req_queue
+        res_queue = non_replicable_dp[0].res_queue
+
+        queue_wrapper = communication.iter.QueueWrapper(
+            communication.protocol.IterDataPipeQueueProtocolClient(req_queue, res_queue)
+        )
+        dispatch_process_dp = communication.iter._IterateQueueDataPipes([queue_wrapper])
```

Comment:
IIUC: 

`IterDataPipeQueueProtocolClient` will be wrapped into a `QueueWrapper` (but still not a `IterDataPipe`), and further wrapped into a `_IterateQueueDataPipes` which is a `IterDataPIpe`?

Reply:
Nope. Both `QueueWrapper` and `_IterateQueueDataPipes` are `IterDataPipe`, this is one of the thing that we can optimize later.